### PR TITLE
Fix CSS split/rejoin

### PR DESCRIPTION
**Pull request by tony19**
This addresses the review comments from #40, including:
 * Change `<style src="...">` (invalid) to `<link rel="stylesheet" href="...">`
 * Add tests

 - [x] CHANGELOG.md has been updated

/cc @hurst @FredKSchott @justinfagnani

### CODE REVIEW

**Comment by tony19** [comment]
The test failures are due to the dom5 bug (which should be fixed by https://github.com/PolymerLabs/dom5/pull/46).

**Review by FredKSchott — COMMENTED** [review]
@tony19 awesome! Thanks for working on this, looks really good.

Lets get that dom5 bug fixed and published and then I'll take a deeper look at the code.

Also adding @justinfagnani since he wrote a lot of this.

**Comment by tony19** [comment]
@FredKSchott Rebased and brought in the dom5 bugfix for integration tests.

**Comment by justinfagnani** [comment]
@tony19 I just published 2.1 of dom5 with the template traversal

**Comment by tony19** [comment]
@justinfagnani Thanks, updated.

**Comment by tony19** [comment]
There's a bug somewhere that prevents CSS rejoining for `build/unbundled`. I'll need to investigate.

File: CHANGELOG.md
Comment (FredKSchott) [inline CHANGELOG.md]: Keep the changelog versions as is on master, we'll make sure it gets merged correctly when we cut the next release.

File: src/polymer-project.ts
Comment (FredKSchott) [inline src/polymer-project.ts]: can you clean up any differences between `splitScript()` & `splitStyle()`? For example, `splitScript` is using a `for()` loop while `splitStyle()` is using `styleTags.forEach()`.

File: src/test/polymer-project_test.ts
Comment (FredKSchott) [inline src/test/polymer-project_test.ts]: `expectedSplitFiles` is always empty at this point, why use push here instead of just setting the array?

**Review by FredKSchott — CHANGES_REQUESTED** [review]
Looks good, just a few questions.

Also, assigning @justinfagnani to review based on his feedback on the original PR.

File: src/polymer-project.ts
Comment (tony19) [inline src/polymer-project.ts]: `splitScript` needs to use a `for`-loop because it uses `continue` at line 236 for unrecognized script types, while `splitStyle` doesn't. I could either (1) refactor this function so that the loop types are the same, or (2) restore the original `for`-loop in `splitStyle` for consistency (i.e., undo cc93d2a57c8f8335246d6d5e179eb4911fcaee37).

File: src/test/polymer-project_test.ts
Comment (tony19) [inline src/test/polymer-project_test.ts]: To maintain its `const`ness. (The same is true for `expectedJoinedFiles[]`)

File: CHANGELOG.md
Comment (tony19) [inline CHANGELOG.md]: Ok, will do. I was anticipating your change in your PR for 0.7.0, which would result in a merge conflict here.

File: src/polymer-project.ts
Comment (FredKSchott) [inline src/polymer-project.ts]: you can use `return` in the `forEach()` iterator to achieve the same behavior as `continue`

File: src/test/polymer-project_test.ts
Comment (FredKSchott) [inline src/test/polymer-project_test.ts]: I like the idea, but this confuses more than it helps. It also doesn't help as much as you would hope, since the array can still be modified.

I'd rather see you use let to simplify, and then Object.freeze() to keep the array from being modified. Or use `setup()` to create the array for each test.

File: src/polymer-project.ts
Comment (tony19) [inline src/polymer-project.ts]: Oh, didn't think of that 👍

File: src/test/polymer-project_test.ts
Comment (tony19) [inline src/test/polymer-project_test.ts]: Ok, will do.

File: CHANGELOG.md
Comment (tony19) [inline CHANGELOG.md]: Done in c552144

File: src/polymer-project.ts
Comment (tony19) [inline src/polymer-project.ts]: Done in 7f1e853

File: src/test/polymer-project_test.ts
Comment (tony19) [inline src/test/polymer-project_test.ts]: Done in 02334e1

**Comment by tony19** [comment]
Let me know if I should squash some of these commits.

**Review by FredKSchott — APPROVED** [review]
LGTM, but we can't merge until we can answer the same concerns that were brought up in https://github.com/Polymer/polymer-build/pull/40. Namely, that we still need to support splitting and not rejoining for CSP compliance (similar to https://github.com/PolymerLabs/crisper). But styles break if they aren't rejoined. So we need to add an option to rejoinHtml() to tell it to only rejoin CSS (and leave scripts split). 

@tony19 do you have any interest in tackling that? Otherwise I can try getting to it early next week.

**Comment by tony19** [comment]
I'm interested, but not sure how soon I could get to it. Can that be a separate PR?

**Review by justinfagnani — CHANGES_REQUESTED** [review]
Please wait for me to finish a review before merging. I still have to think about what it means to split CSS that's not legal externally. We need a way to ensure that `<custom-style>` and styles in templates actually work.

**Comment by stramel** [comment]
Status?

**Comment by cdata** [comment]
> We need a way to ensure that <custom-style> and styles in templates actually work.

This seems like something polymer-cli should ensure, but to venture a proposal: polymer-build should not care. polymer-build is the tool that assumes I know what I'm doing, no?

**Comment by stramel** [comment]
Any update on this? I would love to see this get added in. @justinfagnani

**Comment by abdonrd** [comment]
Any news?

**Comment by stramel** [comment]
I know @rictic worked on updating the CSS parser in analyzer, https://github.com/Polymer/polymer-analyzer/pull/707, maybe he can weigh in on this?

**Comment by enigma127** [comment]
Any news?

**Comment by aomarks** [comment]
Closing this pull request because this repo is now archived. If this PR is still active, please submit a new one at https://github.com/Polymer/tools.

**Pull request closed without merging**
